File: RUST/src/bfs.rs
use std::collections::VecDeque;

// Define a position in the grid as a tuple (row, col)
type Position = (usize, usize);

fn get_index(row: usize, col: usize, cols: usize) -> usize {
    row * cols + col
}

fn is_wall(grid: &Vec<i32>, row: usize, col: usize, cols: usize) -> bool {
    let index = get_index(row, col, cols);
    grid[index] == 1
}

fn bfs(
    grid: &mut Vec<i32>,
    start: Position,
    end: Position,
    rows: usize,
    cols: usize,
) -> (Option<Vec<Position>>, Vec<usize>) {
    // Initialize levels array (distances from start)
    let mut levels = vec![vec![-1; cols]; rows];

    // Initialize visited array
    let mut visited = vec![vec![false; cols]; rows];

    // Track visited nodes in order
    let mut visited_order = Vec::new();

    // Define possible movements (right, left, down, up)
    let dx: [isize; 4] = [1, -1, 0, 0];
    let dy: [isize; 4] = [0, 0, 1, -1];

    // Initialize the queue with start position
    let mut queue = VecDeque::new();
    queue.push_back(start);

    // Mark start as visited
    visited[start.0][start.1] = true;
    levels[start.0][start.1] = 0;
    visited_order.push(get_index(start.0, start.1, cols));

    // Process the queue until it's empty or we've visited the end
    while !queue.is_empty() && !visited[end.0][end.1] {
        // Get next position
        let (curr_x, curr_y) = queue.pop_front().unwrap();

        // Check all four directions
        for i in 0..4 {
            let new_x = curr_x as isize + dy[i];
            let new_y = curr_y as isize + dx[i];

            // Check bounds
            if new_x < 0 || new_x >= rows as isize || new_y < 0 || new_y >= cols as isize {
                continue;
            }

            let new_x = new_x as usize;
            let new_y = new_y as usize;

            // Skip walls and already visited cells
            if is_wall(grid, new_x, new_y, cols) || visited[new_x][new_y] {
                continue;
            }

            // Mark as visited and set level
            visited[new_x][new_y] = true;
            levels[new_x][new_y] = levels[curr_x][curr_y] + 1;

            // Add to visited order
            visited_order.push(get_index(new_x, new_y, cols));

            // Add to queue
            queue.push_back((new_x, new_y));
        }
    }

    // Reconstruct path if end is reachable
    let path = if levels[end.0][end.1] != -1 {
        let mut path = Vec::new();
        let mut curr = end;
        path.push(curr);

        // Work backwards from end to start
        let mut curr_level = levels[curr.0][curr.1];

        while curr != start {
            for i in 0..4 {
                let prev_x = curr.0 as isize + dy[i];
                let prev_y = curr.1 as isize + dx[i];

                // Check bounds
                if prev_x < 0 || prev_x >= rows as isize || prev_y < 0 || prev_y >= cols as isize {
                    continue;
                }

                let prev_x = prev_x as usize;
                let prev_y = prev_y as usize;

                // If this is the previous cell in the path
                if visited[prev_x][prev_y] && levels[prev_x][prev_y] == curr_level - 1 {
                    curr = (prev_x, prev_y);
                    curr_level -= 1;
                    path.push(curr);
                    break;
                }
            }
        }

        // Reverse to get path from start to end
        path.reverse();

        // Mark path cells in the grid
        for &(row, col) in &path {
            let index = get_index(row, col, cols);
            if grid[index] != 1 { // Don't mark walls
                grid[index] = 3; // Mark as path
            }
        }

        Some(path)
    } else {
        None
    };

    // Mark all visited cells in the grid (if they're not already part of the path)
    for &index in &visited_order {
        if grid[index] == 0 {
            grid[index] = 2; // Mark as visited
        }
    }

    (path, visited_order)
}

// Get the indexes of nodes in shortest path
fn get_path_indexes(path: &Vec<Position>, cols: usize) -> Vec<usize> {
    path.iter()
        .map(|&(row, col)| get_index(row, col, cols))
        .collect()
}

pub fn find_shortest_path(
    mut grid: Vec<i32>,
    start_row: usize,
    start_col: usize,
    end_row: usize,
    end_col: usize,
    rows: usize,
    cols: usize,
) -> (Vec<i32>, Vec<usize>, Vec<usize>) {
    let start = (start_row, start_col);
    let end = (end_row, end_col);

    let (path_opt, visited_order) = bfs(&mut grid, start, end, rows, cols);

    // Get path indexes if a path was found
    let path_indexes = match path_opt {
        Some(path) => get_path_indexes(&path, cols),
        None => Vec::new()
    };

    (grid, visited_order, path_indexes)
}

File: Algorithims/BFS.cpp
#include <vector>
#include <queue>
#include <tuple>

using namespace std;

using Position = pair<size_t, size_t>;

size_t get_index(size_t row, size_t col, size_t cols) {
    return row * cols + col;
}

bool is_wall(const vector<int>& grid, size_t row, size_t col, size_t cols) {
    size_t index = get_index(row, col, cols);
    return grid[index] == 1;
}

pair<vector<Position>, vector<size_t>> bfs(
    vector<int>& grid,
    Position start,
    Position end,
    size_t rows,
    size_t cols
) {
    vector<vector<int>> levels(rows, vector<int>(cols, -1));
    vector<vector<bool>> visited(rows, vector<bool>(cols, false));
    vector<size_t> visited_order;

    int dx[4] = {1, -1, 0, 0};
    int dy[4] = {0, 0, 1, -1};

    queue<Position> q;
    q.push(start);
    visited[start.first][start.second] = true;
    levels[start.first][start.second] = 0;
    visited_order.push_back(get_index(start.first, start.second, cols));

    while (!q.empty() && !visited[end.first][end.second]) {
        auto [curr_x, curr_y] = q.front();
        q.pop();

        for (int i = 0; i < 4; ++i) {
            int new_x = static_cast<int>(curr_x) + dy[i];
            int new_y = static_cast<int>(curr_y) + dx[i];

            if (new_x < 0 || new_x >= static_cast<int>(rows) ||
                new_y < 0 || new_y >= static_cast<int>(cols)) {
                continue;
            }

            size_t nx = static_cast<size_t>(new_x);
            size_t ny = static_cast<size_t>(new_y);

            if (is_wall(grid, nx, ny, cols) || visited[nx][ny]) {
                continue;
            }

            visited[nx][ny] = true;
            levels[nx][ny] = levels[curr_x][curr_y] + 1;
            visited_order.push_back(get_index(nx, ny, cols));
            q.emplace(nx, ny);
        }
    }

    vector<Position> path;
    if (levels[end.first][end.second] != -1) {
        Position curr = end;
        path.push_back(curr);
        int curr_level = levels[curr.first][curr.second];

        while (curr != start) {
            for (int i = 0; i < 4; ++i) {
                int prev_x = static_cast<int>(curr.first) + dy[i];
                int prev_y = static_cast<int>(curr.second) + dx[i];

                if (prev_x < 0 || prev_x >= static_cast<int>(rows) ||
                    prev_y < 0 || prev_y >= static_cast<int>(cols)) {
                    continue;
                }

                size_t px = static_cast<size_t>(prev_x);
                size_t py = static_cast<size_t>(prev_y);

                if (visited[px][py] && levels[px][py] == curr_level - 1) {
                    curr = {px, py};
                    curr_level--;
                    path.push_back(curr);
                    break;
                }
            }
        }

        reverse(path.begin(), path.end());

        for (const auto& [r, c] : path) {
            size_t idx = get_index(r, c, cols);
            if (grid[idx] != 1) {
                grid[idx] = 3;
            }
        }
    }

    for (size_t idx : visited_order) {
        if (grid[idx] == 0) {
            grid[idx] = 2;
        }
    }

    return {path, visited_order};
}

vector<size_t> get_path_indexes(const vector<Position>& path, size_t cols) {
    vector<size_t> indexes;
    for (const auto& [row, col] : path) {
        indexes.push_back(get_index(row, col, cols));
    }
    return indexes;
}

tuple<vector<int>, vector<size_t>, vector<size_t>> find_shortest_path(
    vector<int> grid,
    size_t start_row,
    size_t start_col,
    size_t end_row,
    size_t end_col,
    size_t rows,
    size_t cols
) {
    Position start = {start_row, start_col};
    Position end = {end_row, end_col};

    auto [path, visited_order] = bfs(grid, start, end, rows, cols);
    vector<size_t> path_indexes = get_path_indexes(path, cols);

    return {grid, visited_order, path_indexes};
}
